cmd/util/ledger/migrations/deduplicate_contract_names_migration.go

```diff
@@ -0,0 +1,134 @@
+package migrations
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/onflow/cadence/runtime/common"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/fvm/environment"
+	"github.com/onflow/flow-go/fvm/storage/state"
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/convert"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// DeduplicateContractNamesMigration checks if the contract names have been duplicated and
+// removes the duplicate ones.
+type DeduplicateContractNamesMigration struct {
+	log zerolog.Logger
+}
+
+func (d *DeduplicateContractNamesMigration) InitMigration(
+	log zerolog.Logger,
+	_ []*ledger.Payload,
+	_ int,
+) error {
+	d.log = log.
+		With().
+		Str("migration", "DeduplicateContractNamesMigration").
+		Logger()
+
+	return nil
+}
+
+func (d *DeduplicateContractNamesMigration) MigrateAccount(
+	ctx context.Context,
+	address common.Address,
+	payloads []*ledger.Payload,
+) ([]*ledger.Payload, error) {
+	snapshot, err := util.NewPayloadSnapshot(payloads)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create payload snapshot: %w", err)
+	}
+	transactionState := state.NewTransactionState(snapshot, state.DefaultParameters())
+	accounts := environment.NewAccounts(transactionState)
+	flowAddress := flow.ConvertAddress(address)
+
+	contractNames, err := accounts.GetContractNames(flowAddress)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get contract names: %w", err)
+	}
+	if len(contractNames) == 0 {
+		return payloads, nil
+	}
+
+	contractNamesSet := make(map[string]struct{})
+	removeIndexes := make([]int, 0)
+	for i, name := range contractNames {
+		if _, ok := contractNamesSet[name]; ok {
+			// duplicate contract name
+			removeIndexes = append(removeIndexes, i)
+			continue
+		}
+
+		contractNamesSet[name] = struct{}{}
+	}
+
+	if len(removeIndexes) == 0 {
+		return payloads, nil
+	}
+
+	log.Info().
+		Str("address", address.Hex()).
+		Strs("contract_names", contractNames).
+		Msg("removing duplicate contract names")
+
+	// remove the duplicate contract names, keeping the original order
+	for i := len(removeIndexes) - 1; i >= 0; i-- {
+		contractNames = append(contractNames[:removeIndexes[i]], contractNames[removeIndexes[i]+1:]...)
+	}
```

Comment:
Another approach here is to avoid new memory allocations by:
- Sorting `contractNames` in-place (we're not expecting large number of contract names)
- Removing duplicates in-place

```suggestion
	// Sort contactNames in-place
	sort.StringSlice(contractNames).Sort()

	// Remove duplicate in-place
	var foundDuplicate bool
	i := 1
	for i < len(contractNames) {
		if contractNames[i-1] != contractNames[i] {
			i++
			continue
		}
		// Found duplicate (contactNames[i-1] == contactNames[i])
		// Remove contractNames[i]
		copy(contractNames[i:], contractNames[i+1:])
		contractNames = contractNames[:len(contractNames)-1]
		foundDuplicate = true
	}

	if !foundDuplicate {
		return payloads, nil
	}
```